File: chapters/16-zalacznik-i-kod/code/moving_average.cpp
#include <algorithm>
#include <cstddef>
#include <stdexcept>
#include <vector>
std::vector<double> moving_average(
    const std::vector<double>& samples,
    std::size_t window) {
    if (window == 0) {
        throw std::invalid_argument("window must be positive");
    }
    std::vector<double> result;
    result.reserve(samples.size());
    double running_sum = 0.0;
    for (std::size_t index = 0; index < samples.size(); ++index) {
        running_sum += samples[index];
        if (index >= window) {
            running_sum -= samples[index - window];
        }
        const auto count = std::min(index + 1, window);
        result.push_back(running_sum / static_cast<double>(count));
    }
    return result;
}

File: chapters/16-zalacznik-i-kod/code/moving_average.py
"""Moving-average filter used in Appendix G."""

from collections.abc import Sequence


def moving_average(samples: Sequence[float], window: int) -> list[float]:
    """Return a causal moving average with a shortened initial window."""
    if window <= 0:
        raise ValueError("window must be positive")

    result: list[float] = []
    running_sum = 0.0

    for index, sample in enumerate(samples):
        running_sum += sample
        if index >= window:
            running_sum -= samples[index - window]

        current_window = min(index + 1, window)
        result.append(running_sum / current_window)

    return result


if __name__ == "__main__":
    data = [1.0, 1.4, 2.1, 2.8, 3.0]
    print(moving_average(data, window=3))
